### DISCUSSION TITLE
論模組(module)

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS4719-C/Python-Community-GS4719-C#184] [General]
*.py是用Python解釋的腳本(script)。若此*.py是有特別功能的檔案，可稱之為「模組。」很多個模組放在同一個資料夾中，就成為「套件包(package)。」裝這些*.py檔案的資料夾，就稱為「目錄。」

**Comment by EricAgumonJou** [comment]
那我來補充一些好用的內建模組及其功能：

1. NumPy: 是科學應用程式庫的主要軟體包之一，用於處理大型多維陣列和矩陣，它大量的高階數學函式集合和實現方法使得這些物件執行操作成為可能。

2. Pandas:強化了資料處理的方便性也能與處理網頁資料與資料庫資料等，有點類似於Office的Excel能更加方便的進行運算、分析等。
另外，Pandas提供以下三種資料結構供使用：
a. Series：用來處理時間序列相關的資料(如感測器資料等)，主要為建立索引的一維陣列。
b. DataFrame：用來處理結構化的資料，有列索引與欄標籤的二維資料集，例如關聯式資料庫、CSV 等等。
c. Panel：用來處理有資料及索引、列索引與欄標籤的三維資料集(除了特殊需求之外少使用… 略過)。
(資料來源：https://ithelp.ithome.com.tw/articles/10208412)

**Reply by chia-kai-chang** [reply]
說的沒錯，pandas的Series與DataFrame這兩種資料結構是重要概念。

**Reply by Jie2222** [reply]
numpy跟pandas都是用來處理矩陣、陣列，不是很懂甚麼時候要用pandas

**Reply by chia-kai-chang** [reply]
> numpy跟pandas都是用來處理矩陣、陣列，不是很懂甚麼時候要用pandas

@Jie2222 

numpy可以用來處理array，並對array做數學計算或是視為矩陣，做矩陣計算。

pandas是用來把類excel的table資料做『資料清洗』用的，比如填補空缺值，刪除空缺資料。在做資料分析前，『資料分析師』需要先對資料做資料清洗。以方便後續的『資料科學家』使用scikit-learn對資料做機器學習。

**Reply by Jie2222** [reply]
瞭解了，謝謝老師。
想再問個問題，資料分析師跟資料科學家會是兩個獨立的工作 or 職缺嗎

**Reply by chia-kai-chang** [reply]
> 瞭解了，謝謝老師。 想再問個問題，資料分析師跟資料科學家會是兩個獨立的工作 or 職缺嗎

不同公司有不同做法。但是一般來說，「資料分析師」的進入門檻較低(大學或碩班畢業，或是參加過資策會的就業養成班)，使用現成的套件做資料清理、資料分析。「資料科學家」的進入門檻較高(資工碩或博班)，主要為自行設計Python套件、以自定義的損失函數開發新的演算法。

**Reply by Jie2222** [reply]
了解，謝謝老師。

**Comment by EricAgumonJou** [comment]
「機器學習」常用的模組如：

3. Scikit-Learn ：是 Python 使用者入門機器學習的一個高階、設計成熟且友善的套件模組，其建構於 NumPy、SciPy 與 Matplotlib，是開源並可作為商業使用的套件模組。它為許多標準的機器學習和資料探勘任務提供演算法，如聚類、迴歸、分類、降維和模型選擇。其應用場景可以被簡單分類為：

預處理（Preprocessing）
監督式學習（Supervised learning）
分類（Classification）
迴歸（Regression）
非監督式學習（Unsupervised learning）
分群（Clustering）
降維（Dimensionality reduction）
模型選擇（Model selection）
預處理的功能呼應了資料科學專案中的整併以及轉換；「監督式學習」、「非監督式學習」與模型選擇的功能則呼應了專案中的預測。
（資料來源：https://yaojenkuo.io/ml-newbies/04-sklearn.html）

**Reply by chia-kai-chang** [reply]
說得很正確

**Comment by EricAgumonJou** [comment]
「深度學習」常用到的模組：

4. TensorFlow: 是由 Google Brain 開發的深度學習架構。它提供使用具有多個資料集的「人工類神經網路（Neural Networks, NN）」的能力。在最流行的 TensorFlow應用中有目標識別、語音識別等。TensorFlow亦用於各種感知和語言理解任務的機器學習任務上。 可應用於研究和生產Google商業產品，如語音辨識、Gmail、Google 相冊和搜索，其中許多產品曾使用過其前代軟體DistBelief。
（資料來源：https://tensorflow.google.cn/about）

**Reply by chia-kai-chang** [reply]
沒錯

**Comment by chia-kai-chang** [comment]
> _.py是用Python解釋的腳本(script)。若此_.py是有特別功能的檔案，可稱之為「模組。」很多個模組放在同一個資料夾中，就成為「套件包(package)。」裝這些*.py檔案的資料夾，就稱為「目錄。」

裝任何東西的**資料夾**都是**目錄**唷~

資料夾跟目錄是相同的。

這邊應該要強調，如果一個資料夾裡面有很多*.py的檔案(模組)，這個資料夾就可稱為package(程式包)

如：`C:\Program Files\Python39\Lib\site-packages\matplotlib`
![image](https://user-images.githubusercontent.com/1768891/145194672-ae293821-5e4c-4676-84e1-6f190a864570.png)

**Reply by enoch-lightning** [reply]
@benctw 老師，我搞不太清楚：C:\Program Files\Python39\Lib\site-packages\matplotlib與圖片中的
本機>本機磁碟(C:)>Program Files>Python39>Lib>site-packages>matplotlib是同一個東西嗎？謝謝老師。

**Reply by chia-kai-chang** [reply]
是同一個東西沒錯

**Reply by enoch-lightning** [reply]
收到，謝謝老師。

**Reply by enoch-lightning** [reply]
我現在才突然意會到，老師想表達的意思，就是matplolib這個「資料夾=目錄」裡有很多`*.py`(`*.py`就一定是「模組(module)」，「模組」就是`*.py`的別名)的檔案，如
`__init__.py`
`_animation_data.py`
`_c_internal_utils.cp39-win_amd64.pyd`
`_cm.py`
`_cm_listed.py`
`_color_data.py`
`_constrained_layout.py`
`_contour.cp39-win_amd64.py`
`_enums.py`
`_image.cp39-win_amd64.pyd`
`_internal_utils.py`
`_layoutgrid.py`
`_mathtext.py`
`_path.cp39-win_amd64.pyd`
所以matplotlib是一個「程式包(英文是package)」。
我這樣理解對嗎？

**Reply by enoch-lightning** [reply]
經過今日與老師討論，老師說：對。
在討論過程中，了解到package的中文可以叫「程式包」，也可以叫作「套件。」

**Reply by enoch-lightning** [reply]
我發現我的電腦下載matplotlib的目錄(資料夾)(還是「路徑」(path)？)與老師不一樣，我下載matplotlib的目錄是C:\Program Files\Python39\Lib\site-packages\matplotlib
這個目錄如果截圖下來會長下面這樣
![image](https://user-images.githubusercontent.com/86505454/151930371-ea9aa31e-a3c2-45fd-b770-7e7f4afd3634.png)
C:\Program Files\Python39\Lib\site-packages\matplotlib就是圖中的「使用者>enoch>AppData>Local>Programs>Python>Python39>Lib>site-packages>matplotlib」

**Comment by ghost** [comment]
這篇的留言好豐富，ai跨領域學分學程有一門機器學習概論就有教前面有其他同學提到的pandas和numpy在資料處理要如何使用，有興趣的可以修看看!

**Reply by 109802503gloria** [reply]
機器學習概論在有了一些python的基礎知識後，會比較理解老師的程式碼在做什麼，期末的專題也有可能會用到flask，可以學到很多東西。
